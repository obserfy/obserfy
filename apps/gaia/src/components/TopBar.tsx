import { Menu, Transition } from "@headlessui/react"
import { useRouter } from "next/router"
import { FC, Fragment } from "react"
import clsx from "clsx"
import Button from "$components/Button/Button"
import AlertDialog from "$components/AlertDialog"
import useToggle from "$hooks/useToggle"
import Icon from "$components/Icon/Icon"
import StudentSelector from "$components/StudentSelector"
import useGetChild from "$hooks/api/useGetChild"
import useGetChildren from "$hooks/api/useGetChildren"
import useGetUser from "$hooks/api/useGetUser"
import { useQueryString } from "$hooks/useQueryString"

const TopBar = () => {
  const router = useRouter()
  const studentId = useQueryString("studentId")
  const { data: student } = useGetChild(studentId)
  const { data: user } = useGetUser()
  const { data: students } = useGetChildren()

  return (
    <div className="relative sm:sticky sm:top-0 z-20 h-16 bg-surface bg-opacity-90 sm:border-b backdrop-filter backdrop-blur-lg">
      <div className="flex items-center px-4 sm:pl-0 h-16">
        <div className="hidden sm:block flex-shrink-0 px-4 mr-4 w-sidebar font-bold text-gray-600 border-r">
          {student?.schoolName}
        </div>

        {students && student && (
          <StudentSelector
            students={students}
            selectedStudent={student}
            setSelectedStudent={(s) => {
              router.push(router.asPath.replace(student.id, s.id))
            }}
          />
        )}

        <UserOptions name={user?.name} picture={user?.picture} />
      </div>
    </div>
  )
}

const UserOptions: FC<{
  name: string | undefined
  picture: string | undefined
}> = ({ picture, name }) => {
  const logout = useToggle()

  return (
    <>
      <LogoutAlertDialog isOpen={logout.isOn} close={logout.toggle} />
      <Menu as="div" className="inline-block relative ml-auto text-left">
        <Menu.Button className="group flex flex-shrink-0 items-center rounded-full">
          {/* eslint-disable-next-line @next/next/no-img-element */}
          <img src={picture} className="w-8 h-8 rounded-full" alt="" />

          <p className="hidden md:block ml-2 text-sm font-semibold text-gray-700 group-hover:text-primary-800">
            {name}
          </p>

          <Icon
            src="/icons/chevron-down.svg"
            className="hidden sm:block !w-6 !h-6"
            color="bg-gray-800 group-hover:bg-primary-800"
          />
        </Menu.Button>

        <Transition
          as={Fragment}
          enter="transition ease-out duration-100"
          enterFrom="transform opacity-0 scale-95"
          enterTo="transform opacity-100 scale-100"
          leave="transition ease-in duration-75"
          leaveFrom="transform opacity-100 scale-100"
          leaveTo="transform opacity-0 scale-95"
        >
          <Menu.Items className="absolute right-0 mt-2 w-40 bg-white rounded-md ring-1 ring-black ring-opacity-5 shadow-lg origin-top-right focus:outline-none">
            <div className="py-1">
              <Menu.Item>
                {({ active }) => (
                  <button
                    onClick={logout.toggle}
                    className={clsx(
                      active ? "text-gray-900 bg-gray-100" : "text-gray-700",
                      "group flex items-center py-2 px-4 w-full text-sm text-left"
                    )}
                  >
                    <Icon
                      src="/icons/log-out.svg"
                      className="mr-3 !w-4 !h-4"
                      color="bg-gray-700 group-hover:text-primary-700"
                    />
                    Sign out
                  </button>
                )}
              </Menu.Item>
            </div>
          </Menu.Items>
        </Transition>
      </Menu>
    </>
  )
}

const LogoutAlertDialog: FC<{
  isOpen: boolean
  close: () => void
}> = ({ close, isOpen }) => (
  <AlertDialog
    isOpen={isOpen}
    close={close}
    title="Sign Out"
    description="Do you want to sign out from Obserfy?"
    actionButton={
      // eslint-disable-next-line @next/next/no-html-link-for-pages
      <a href="/api/auth/logout">
        <Button className="sm:ml-3 w-full sm:w-auto text-base sm:text-sm hover:!bg-red-700 focus:!ring-red-500 !bg-red-600">
          Sign Out
        </Button>
      </a>
    }
    icon={
      <div className="flex flex-shrink-0 justify-center items-center mx-auto sm:mx-0 w-12 sm:w-10 h-12 sm:h-10 bg-red-100 rounded-full">
        <Icon
          src="/icons/log-out.svg"
          className="w-6 h-6"
          color="bg-red-600"
          aria-hidden="true"
        />
      </div>
    }
  />
)

export default TopBar
