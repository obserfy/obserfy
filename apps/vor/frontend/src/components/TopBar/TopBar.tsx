import { FC } from "react"
import { Flex, ThemeUIStyleObject } from "theme-ui"
import { Trans } from "@lingui/macro"
import BackButton from "../BackButton/BackButton"
import Breadcrumb from "../Breadcrumb/Breadcrumb"
import BreadcrumbItem from "../Breadcrumb/BreadcrumbItem"

export interface BreadcrumbData {
  to?: string
  text: string
}
export interface TopBarProps {
  breadcrumbs: BreadcrumbData[]
  containerSx?: ThemeUIStyleObject
}
export const TopBar: FC<TopBarProps> = ({ breadcrumbs, containerSx }) => (
  <Flex sx={{ height: 48, alignItems: "center", ...containerSx }}>
    <BackButton
      to={breadcrumbs[Math.max(breadcrumbs.length - 2, 0)].to ?? "/"}
      state={{ preserveScroll: true }}
    />
    <Breadcrumb>
      {breadcrumbs.map(({ to, text }) => (
        <BreadcrumbItem key={text + (to ?? "empty")} to={to}>
          <Trans id={text} />
        </BreadcrumbItem>
      ))}
    </Breadcrumb>
  </Flex>
)

export const breadCrumb = (text?: string, to?: string): BreadcrumbData => ({
  text: text ?? "",
  to,
})

export default TopBar
