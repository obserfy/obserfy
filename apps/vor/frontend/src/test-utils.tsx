/* eslint-disable import/no-extraneous-dependencies */
import { FC, ReactElement } from "react"
import { QueryClient, QueryClientProvider } from "react-query"
import { ThemeProvider, Box, Theme } from "theme-ui"
import { render, RenderOptions, RenderResult } from "@testing-library/react"
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router"
import { I18nProvider } from "@lingui/react"
import { i18n } from "@lingui/core"
import { en } from "make-plural/plurals"
import theme from "./gatsby-plugin-theme-ui"
// @ts-ignore
import enCatalog from "../i18n/lingui/en/messages.js"

// any used here because the catalogs are generated b lingui
// not our responsibility
i18n.loadLocaleData("en", { plurals: en })
i18n.load("en", enCatalog.messages as any)
i18n.activate("en")

const queryClient = new QueryClient()
const history = createHistory(createMemorySource("/"))

const AllProviders: FC = ({ children }) => {
  return (
    <I18nProvider i18n={i18n}>
      <QueryClientProvider client={queryClient}>
        <LocationProvider history={history}>
          <ThemeProvider theme={theme as Theme}>
            <Box sx={{ fontSize: [16, 20] }}>{children}</Box>
          </ThemeProvider>
        </LocationProvider>
      </QueryClientProvider>
    </I18nProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: AllProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
