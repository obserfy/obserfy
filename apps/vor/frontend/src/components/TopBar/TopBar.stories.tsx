import { Meta, Story } from "@storybook/react"
import TopBar, { TopBarProps } from "./TopBar"

export default {
  title: "Core/TopBar",
  component: TopBar,
  parameters: {
    componentSubtitle: "Just a simple TopBar",
  },
} as Meta

const Template: Story<TopBarProps> = ({ breadcrumbs }) => (
  <TopBar breadcrumbs={breadcrumbs} />
)

export const Default = Template.bind({})
Default.args = {
  breadcrumbs: [{ to: "/", text: "oblivion" }],
}
