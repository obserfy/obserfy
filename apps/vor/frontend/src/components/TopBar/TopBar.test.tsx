import { render } from "../../test-utils"
import { Default } from "./TopBar.stories"

describe("TopBar", () => {
  it("should render correctly", () => {
    const { container } = render(
      <Default breadcrumbs={[{ to: "/", text: "test" }]} />
    )
    expect(container).toMatchSnapshot()
  })
})
