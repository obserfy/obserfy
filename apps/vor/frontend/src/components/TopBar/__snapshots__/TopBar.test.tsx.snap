// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar should render correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 16px;
}

@media screen and (min-width: 40em) {
  .emotion-0 {
    font-size: 20px;
  }
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 48px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  margin: 8px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  display: inline-block;
  text-align: center;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 8px;
  padding-bottom: 8px;
  color: white;
  background-color: var(--theme-ui-colors-primary);
  border: 0;
  border-radius: 4px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  box-sizing: border-box;
  font-size: 14px;
  color: var(--theme-ui-colors-textPrimary);
  background-color: transparent;
  border-radius: 6px;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  -webkit-transition: background-color 125ms cubic-bezier(0.0, 0.0, 0.2, 1);
  transition: background-color 125ms cubic-bezier(0.0, 0.0, 0.2, 1);
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 16px;
  padding-right: 16px;
  white-space: nowrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  touch-action: manipulation;
  box-shadow: none;
  border-color: var(--theme-ui-colors-border);
  border-style: solid;
  border-width: 1px;
  line-height: 1;
  border: none;
  padding: 4px;
}

.emotion-3:disabled {
  opacity: 0.38;
  cursor: auto;
}

.emotion-3:hover,
.emotion-3:focus {
  background-color: var(--theme-ui-colors-primaryLighter);
  border-color: var(--theme-ui-colors-primary);
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  fill: var(--theme-ui-colors-icon);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 20px;
  height: 20px;
  fill: var(--theme-ui-colors-textMediumEmphasis);
}

.emotion-5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow-x: auto;
}

.emotion-6 {
  display: inline-block;
  font-size: 13px;
  color: var(--theme-ui-colors-textMediumEmphasis);
  white-space: nowrap;
}

.emotion-6:last-child {
  color: var(--theme-ui-colors-textPrimary);
}

.emotion-6:last-child>.splitter {
  display: none;
}

.emotion-7:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-8 {
  margin-left: 4px;
  margin-right: 4px;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <a
        aria-current="page"
        class="emotion-2"
        data-cy="back-button"
        href="/"
      >
        <button
          class="emotion-3"
        >
          <div
            class="emotion-4"
          />
        </button>
      </a>
      <div
        class="emotion-5"
      >
        <a
          aria-current="page"
          class="emotion-6"
          href="/"
        >
          <span
            class="emotion-7"
          >
            test
          </span>
          <span
            class="splitter emotion-8"
          >
            /
          </span>
        </a>
      </div>
    </div>
  </div>
</div>
`;
